### DISCUSSION TITLE
What defines a plugin vs a command?

### DISCUSSION THREAD

**Discussion by greenaar** [woodpecker-ci/woodpecker#4375] [Ask a Question] [answered]
Apologies for what is probably a very simple question.

So "`settings:` is reserved for use with plugins, and is incompatible with `commands:`" -- is there something intrinsic that says 'this is a plugin', or is it just a case of there's a `settings:` block, so it's therefore a plugin?

On the same track, is `environment` limited to commands only, as `settings:` looks to effectively be environment variables + prefix?

Are there any other combinations? I know `secrets:` looks to be on the way out (the replacement method is easy enough to use).

Thanks for your patience.

**Comment by qwerty287** [comment, marked as answer]
If you use any one these keys:

- `commands`
- `environment`
- `entrypoint`
- `secrets` (deprecated)

it's a command step, otherwise a plugin.

**Comment by zc-devs** [comment]
> is there something intrinsic that says 'this is a plugin

#4343 :)
